database/src/sqlite3.cpp

```diff
@@ -0,0 +1,121 @@
+                                        types::Typedef("sql", ValueType::String) } } } },
+                args);
+
+        sqlite3* db = (sqlite3*)args[0].usertypeRef().data();
+        const char* sql = args[1].string().c_str();
+        char* error_message = nullptr;
+
+        int code = sqlite3_exec(db, sql, nullptr, nullptr, &error_message);
+        if (code != SQLITE_OK)
+        {
+            std::runtime_error error = std::runtime_error(
+                "database:sqlite:exec: exited with code " + std::to_string(code) + "\n" + std::string(error_message));
+            sqlite3_free(error_message);
+            throw error;
+        }
+
+        return Ark::Value(code);
+    }
+
+    Value database_sqlite_exec_with_callback(std::vector<Value>& args, Ark::VM* vm [[maybe_unused]])
+    {
+        if (args.size() != 3 || !args[0].usertype().is<sqlite3>() || args[1].valueType() != ValueType::String || !args[2].isFunction())
+            types::generateError(
+                "database:sqlite:exec_with_callback",
+                { { types::Contract { { types::Typedef("database", ValueType::User),
+                                        types::Typedef("sql", ValueType::String),
+                                        types::Typedef("callback", ValueType::Closure) } } } },
+                args);
+
+        sqlite3* db = (sqlite3*)args[0].usertypeRef().data();
+        const char* sql = args[1].string().c_str();
+        const Ark::Value user_defined_callback = args[2];
+
+        // 'sqlite3_exec' accepts a callback function which we will define here. This callback is invoked for
+        // each result row coming out of the evaluated SQL statement(s). The first argument is a void*
+        // where we can pass a pointer to a vector containing all the results.
+        auto callback = [](void* all_results, int argc, char** argv, char** col_names [[maybe_unused]]) -> int {
+            auto all_results_casted = static_cast<std::vector<std::vector<Ark::Value>>*>(all_results);
+            std::vector<Ark::Value> current_row;
+
+            // Add each element as an Ark String to the current row results
+            for (int i = 0; i < argc; i++)
+                current_row.push_back(Ark::Value(std::string(argv[i])));
+
+            // Add the current row to our collection of results
+            all_results_casted->push_back(current_row);
+            return 0;
+        };
+
+        char* error_message = nullptr;
+        std::vector<std::vector<Ark::Value>> all_results;
+        int code = sqlite3_exec(db, sql, callback, &all_results, &error_message);
+
+        // For each row returned from the SQL query, run the user's callback
+        // function on each row passing in an Ark List of the row results.
+        for (std::vector<Ark::Value> row : all_results)
+            vm->resolve(&user_defined_callback, Ark::Value(std::vector<Ark::Value>(row.begin(), row.end())));
```

Comment:
As you're not modifying row, you could make it const& and give it as is to the Value constructor, there should be no need to create another vector from it.

Reply:
Initially when writing this, I played around with passing it as a const reference, but I couldn't exactly get it to work.

Maybe I'm missing something, but the Value constructor only takes in a const value if it's an Ark::Value, but we're trying to pass a `std::vector<Ark::Value>` and the Value constructor can only takes an rvalue of `std::vector<Ark::Value>`.

Perhaps I'm misunderstanding your suggestion. I'm still learning C++, so any tips or guidance would be helpful. Feel free to message me on Discord as well.

Reply:
That could be because the Ark::Value is missing a constructor for this. If you aren't reusing all_result later on, I assume you could use an rvalue (`std::vector<..>&&`) and move it in (`Value(std::move(row))`). Haven't tested that, and this is just me nitpicking ; overall the code is of pretty good quality!

Reply:
Thanks!

I just read up on `std::move`. That's a really cool function. I had no clue about it. I was able to just make `row` a reference and pass it to `std::move`.

Reply:
Nice! std::move takes a *rvalue reference*, row should be of type `std::vector<Value>&&` (mind the double `&`). C++ is kinda weird but this should do the trick.

Don't mind the failing tests in the CI, the console tests are broken (#54).